lib/Dialect/TTNN/Transforms/Workarounds/Decomposition/LinearOpRewritePattern.cpp

```diff
@@ -0,0 +1,151 @@
+
+  // Rank difference will be used to align batch dimensions
+  int64_t outRank = std::max(rankA, rankB) - (rankA == 1 || rankB == 1);
+  int64_t rankDifference = std::max<int64_t>(0, outRank - rankA);
+
+  // Initialize output shape based on the tensor with higher rank
+  SmallVector<int64_t> outputShape;
+  if (rankB > rankA) {
+    outputShape.assign(shapeB.begin(), shapeB.end());
+  } else {
+    outputShape.assign(shapeA.begin(), shapeA.end());
+  }
+
+  // Handle batch dimensions for the case where rankB > rankA
+  for (int64_t index = 0; index < rankDifference; ++index) {
+    // tt-metal requires these front dimensions to be 1
+    // For our purposes, we'll just copy them
+    outputShape[index] = shapeB[index];
+  }
+
+  // Copy dimensions from shapeA except the last one
+  for (int64_t index = 0; index < rankA - 1; ++index) {
+    outputShape[rankDifference + index] = shapeA[index];
+  }
+
+  // The last dimension comes from input_tensor_b
+  outputShape[outputShape.size() - 1] = shapeB[shapeB.size() - 1];
+
+  // Handle the vector matmul case: if rankA == 1, remove the second-to-last
+  // dimension
+  if (rankA == 1 && outputShape.size() > 1) {
+    SmallVector<int64_t> newShape;
+    newShape.reserve(outputShape.size() - 1);
+    // Copy all elements except the second-to-last dimension
+    for (size_t srcIdx = 0; srcIdx < outputShape.size(); ++srcIdx) {
+      if (srcIdx != outputShape.size() - 2) {
+        newShape.push_back(outputShape[srcIdx]);
+      }
+    }
+    outputShape = std::move(newShape);
+  }
+
+  // Handle the case where rankB == 1, remove the last dimension
+  if (rankB == 1) {
+    SmallVector<int64_t> newShape;
+    newShape.reserve(outputShape.size() - 1);
+    for (size_t index = 0; index < outputShape.size() - 1; ++index) {
+      newShape.push_back(outputShape[index]);
+    }
+    outputShape = std::move(newShape);
+  }
+
+  return outputShape;
+}
+
+LogicalResult
+LinearOpRewritePattern::matchAndRewrite(ttnn::LinearOp srcOp,
+                                        PatternRewriter &rewriter) const {
+
+  // Only decompose if input B is batched AND bias exists
+  if (!isBatchedLinearOp(srcOp) || !srcOp.getBias()) {
+    return failure();
+  }
+
+  RankedTensorType inputAType =
+      mlir::cast<RankedTensorType>(srcOp.getA().getType());
+  RankedTensorType inputBType =
+      mlir::cast<RankedTensorType>(srcOp.getB().getType());
+  RankedTensorType outputType =
+      mlir::cast<RankedTensorType>(srcOp.getResult().getType());
+
+  // Compute matmul output shape
+  SmallVector<int64_t> matmulShape =
+      computeMatmulOutputShape(inputAType.getShape(), inputBType.getShape(),
+                               srcOp.getTransposeA(), srcOp.getTransposeB());
+
+  // Create matmul output type
+  auto outputEncoding =
+      mlir::cast<ttnn::TTNNLayoutAttr>(outputType.getEncoding());
+  auto matmulOutputType = RankedTensorType::get(
+      matmulShape, outputType.getElementType(), outputType.getEncoding());
```

Comment:
```suggestion
  auto matmulOutputType = utils::RankedTensorTypeFactory::create(outputType, matmulShape);
```
It's not just nit, encoding can be dependent on the shape, `RankedTensorTypeFactory` handles it gracefully.